extension/jaegerremotesampling/internal/remote_strategy_cache_test.go

```diff
@@ -0,0 +1,247 @@
+
+package internal
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/fortytw2/leaktest"
+	"github.com/jaegertracing/jaeger/thrift-gen/sampling"
+	"github.com/stretchr/testify/assert"
+)
+
+const cacheTestItemTTL = 50 * time.Millisecond
+
+var testStrategyResponseA = &sampling.SamplingStrategyResponse{
+	ProbabilisticSampling: &sampling.ProbabilisticSamplingStrategy{
+		SamplingRate: 0.1337,
+	},
+}
+
+var testStrategyResponseB = &sampling.SamplingStrategyResponse{
+	ProbabilisticSampling: &sampling.ProbabilisticSamplingStrategy{
+		SamplingRate: 0.1337,
+	},
+}
+
+func Test_serviceStrategyCache_ReadWriteSequence(t *testing.T) {
+	cache := newServiceStrategyCache(cacheTestItemTTL).(*serviceStrategyTTLCache)
+	defer func() {
+		assert.NoError(t, cache.close())
+	}()
+
+	// override clock with a test-controlled one
+	testTime := time.Unix(1690221207, 0)
+	cache.clockFn = func() time.Time {
+		return testTime
+	}
+
+	// initial read returns nothing
+	result, ok := cache.get("fooSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+	result, ok = cache.get("barSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	// perform a write for fooSvc at testTime
+	firstWriteTime := testTime
+	cache.put("fooSvc", testStrategyResponseA)
+
+	// whitebox assert for internal timestamp tracking (we don't want a caching bug manifesting as stale data serving)
+	// (post-write)
+	assert.Equal(t, serviceStrategyCacheEntry{
+		retrievedAt:      testTime,
+		strategyResponse: testStrategyResponseA,
+	}, cache.items["fooSvc"])
+
+	// read without time advancing
+	result, ok = cache.get("fooSvc")
+	assert.True(t, ok)
+	assert.Equal(t, testStrategyResponseA, result)
+	result, ok = cache.get("barSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	// reading does not mutate internal cache state
+	assert.Equal(t, serviceStrategyCacheEntry{
+		retrievedAt:      firstWriteTime,
+		strategyResponse: testStrategyResponseA,
+	}, cache.items["fooSvc"])
+
+	// advance time (still within TTL time range)
+	testTime = testTime.Add(20 * time.Millisecond)
+
+	// the written item is still available
+	result, ok = cache.get("fooSvc")
+	assert.True(t, ok)
+	assert.Equal(t, testStrategyResponseA, result)
+	result, ok = cache.get("barSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	// advance time (just before end of TTL time range)
+	testTime = testTime.Add(30 * time.Millisecond)
+
+	// the written item is still available
+
+	result, ok = cache.get("fooSvc")
+	assert.True(t, ok)
+	assert.Equal(t, testStrategyResponseA, result)
+	result, ok = cache.get("barSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	// advance time (across TTL range)
+	testTime = testTime.Add(1 * time.Millisecond)
+
+	// the (now stale) cached item is no longer available
+	result, ok = cache.get("fooSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+	result, ok = cache.get("barSvc")
+	assert.False(t, ok)
+	assert.Nil(t, result)
+
+	assert.Equal(t, serviceStrategyCacheEntry{
```

Comment:
Interesting: this means that the item is still in the cache, it's just not returned when `get` is called.

Reply:
Yeah, I could also just delete stale items on stale reads. I prefer to just let cache misses replace them.
This is sort of why I created a "cleaner" in a periodic goroutine to ensure that the memory footprint doesn't grow unbounded, rather than 100% relying on queries managing to clean up stale items (because queries are for a single service).